### DISCUSSION TITLE
Solver which Handles Additional Constraints

### DISCUSSION THREAD

**Discussion by dsigler1234** [JuliaManifolds/Manopt.jl#162] [Q&A] [answered]
Hi,

I was wondering if any of the solvers in Manopt.jl can handle optimization on a manifold with additional constraints on the variables present. 

See this paper for example: https://arxiv.org/pdf/1901.10000.pdf 

Best,

Devon

**Comment by kellertuer** [comment, marked as answer]
I supervised a master thesis which finished beginning of October implementing both algorithms in the Manopt framework;
so they are planned and will come hopefully as a PR End of November, maybe only mid of December, since I am a little busy with teaching these days.

_edit:_ So the code in general is there, but a little bit of documentation, a little bit of code optimisation / speedup is missing and test coverage. The last post will take a little bit of time I think.
Currently I am also not aware that any other manifold optimisation toolbox (e.g. in Python or matlab) covers these until now.

**Comment by kellertuer** [comment]
Oh, actually one of the solvers _can_ do that but that one is maybe a little tricky, because the subproblem might be he challenging part (and is a constraint problem itself) see https://manoptjl.org/stable/solvers/FrankWolfe/ based on the paper https://arxiv.org/abs/1710.10770

**Comment by kellertuer** [comment]
Update, I at least started n #163 introducing this, but the fine tuning, that is tests/docs/examples are still to be checked.

**Comment by kellertuer** [comment]
Sincethis morning: https://manoptjl.org/stable/solvers/augmented_Lagrangian_method/ – https://manoptjl.org/stable/solvers/exact_penalty_method/.

I have a tutorial planned, but since Pluto notebooks work standalone, the new version has to have the algorithms first :)

**Reply by dsigler1234** [reply]
Ok sounds good. Thanks for keeping me updated. I appreciate it. 

 Just some background on me.  My name is Devon Sigler I am researcher at the national renewable energy lab (NREL). We are interested in using manifold optimization to solve Optimal AC power flow problems. Over the next two years we have some funding to explore this. Of course to do AC powerlfow we need methods that can handle a few constraints on the manifold. This was the reason for my question. We are interested in bench marking several manifold optimization methods on a set to AC test systems we have. Hence, we are interested in using any methods added to this code base which can handle constraints. I would also be interested in potentially contributing some new methods we might develop.

**Reply by mateuszbaran** [reply]
That's a cool application. Could you share an example problem (objective function and constraints) we could use to benchmark Manopt.jl? This would help us make Manopt.jl faster on problems that are practically relevant.

**Reply by dsigler1234** [reply]
Hi Sorry for my delayed response. I got pulled away to another project for a few weeks.  As far as the problem we want to solve, you can find it here https://lanl-ansi.github.io/PowerModels.jl/stable/math-model/. We are looking at the first AC Optimal  Power Flow Problem presented. Equations 8,9,10 are the manifold we want to stay on, while considering the other constraints (equation (5) can be ignored). Since variables are complex valued, we usually represent some variables in polar form and others in rectangular form. I believe equations (8)-(10) can be an embedded manifold in Euclidean space, much like the torus example in Manifolds.jl (https://github.com/JuliaManifolds/Manifolds.jl/blob/master/src/manifolds/EmbeddedTorus.jl).  However, I believe I will need to implement my own custom manifold for this application. Do you have any recommendations for doing that?

**Reply by kellertuer** [reply]
Sounds cool – we actually have a tutorial how to implement a manifold at https://juliamanifolds.github.io/ManifoldsBase.jl/stable/example.html – let me know if there are any questions :)

**Reply by mateuszbaran** [reply]
Thanks for the link. These three equations indeed look like they define a manifold but there is a small problem. Riemannian optimization methods rely on fast calculation of geodesics (at least geodesics; more complex methods need additional tools). I think in general your constraints would require a generic ODE-based geodesic solver. `EmbeddedTorus` does that using explicit parametrizations which are switched as necessary. 

Manopt isn't designed to work well for such problems but who knows, it could likely be adapted to handle it well. If you don't see how to make other approaches work well enough then it may be worth a try. I can help you with this work.

**Comment by kellertuer** [comment]
Cool. For now I do not have more methods nor am I aware of more such methods, but I am always interested in extending the library of algorithms of course. And if the code is used and cited, I am of course also always happy :)
